import "yt/yt_proto/yt/formats/extension.proto";

package NYT.NTesting;

message TUrlRow {
    optional string host = 1     [(NYT.column_name) = "Host"];
    optional string path = 2     [(NYT.column_name) = "Path"];
    optional sint32 http_code = 3 [(NYT.column_name) = "HttpCode"];
}

message TGoodUrl {
    optional string Url = 1     [(NYT.column_name) = "Url"];
}

message THostRow {
    optional string host = 1          [(NYT.column_name) = "Host"];
    optional sint32 http_code_total = 2 [(NYT.column_name) = "HttpCodeTotal"];
}

message TRowVer1 {
    optional string string_1 = 1;
    optional uint32 uint32_2 = 2;
}

message TRowVer2 {
    optional string string_1 = 1;
    optional uint32 uint32_2 = 2;
    optional fixed64 fixed64_3 = 3;
}

message TRowFieldSerializationOption {
    optional TUrlRow url_row_1 = 1 [(NYT.flags) = SERIALIZATION_YT];
    optional TUrlRow url_row_2 = 2;
}

message TRowMessageSerializationOption {
    option (NYT.default_field_flags) = SERIALIZATION_YT;
    optional TUrlRow url_row_1 = 1;
    optional TUrlRow url_row_2 = 2;
}

message TRowMixedSerializationOptions {
    option (NYT.default_field_flags) = SERIALIZATION_YT;
    optional TUrlRow url_row_1 = 1;
    optional TUrlRow url_row_2 = 2 [(NYT.flags) = SERIALIZATION_PROTOBUF];
}

message TRowSerializedRepeatedFields {
    option (NYT.default_field_flags) = SERIALIZATION_YT;
    repeated int64 ints = 1;
    repeated int64 packed_ints = 2 [packed=true];
    repeated TUrlRow url_rows = 3;
}

message TUrlRowWithColumnNames {
    optional string host = 1     [(NYT.column_name) = "Host_ColumnName", (NYT.key_column_name) = "Host_KeyColumnName"];
    optional string path = 2     [(NYT.key_column_name) = "Path_KeyColumnName"];
    optional sint32 http_code = 3;
}

message TRowMixedSerializationOptions_ColumnNames {
    option (NYT.default_field_flags) = SERIALIZATION_YT;
    optional TUrlRowWithColumnNames url_row_1 = 1;
    optional TUrlRowWithColumnNames url_row_2 = 2 [(NYT.flags) = SERIALIZATION_PROTOBUF];
}

message TBadProtobufSerializedRow {
    // Repeated fields are not supported in PROTOBUF serialization mode.
    repeated int64 ints = 1;
}

message TRowWithTypeOptions
{
    enum Color
    {
        WHITE = 0;
        BLUE = 1;
        RED = -1;
    }

    message Embedded
    {
        optional Color color_int_field = 1 [(NYT.flags) = ENUM_INT];
        optional Color color_string_field = 2 [(NYT.flags) = ENUM_STRING];
        optional bytes any_field = 3 [(NYT.flags) = ANY];
    }

    optional Color color_string_field = 2 [(NYT.flags) = ENUM_STRING];
    optional bytes other_columns_field = 4 [(NYT.flags) = OTHER_COLUMNS];
    optional Color color_int_field = 1 [(NYT.flags) = ENUM_INT];
    optional Embedded embedded_field = 5 [(NYT.flags) = SERIALIZATION_YT];
    repeated Color repeated_enum_int_field = 6 [(NYT.flags) = SERIALIZATION_YT, (NYT.flags) = ENUM_INT];
    optional bytes any_field = 3 [(NYT.flags) = ANY];
}

message TGrepperRecord
{
    optional string key = 1;
    optional bytes other = 2 [(NYT.flags) = OTHER_COLUMNS];
}

message TNumberRecord
{
    optional int64 Number = 1 [(NYT.column_name) = "number"];
}
